tensorboard/data/server/server.rs

```diff
@@ -0,0 +1,216 @@
+        res.runs.push(data::Run {
+            name: "train".to_string(),
+            start_time: FAKE_START_TIME,
+            ..Default::default()
+        });
+        Ok(Response::new(res))
+    }
+
+    async fn list_scalars(
+        &self,
+        _request: Request<data::ListScalarsRequest>,
+    ) -> Result<Response<data::ListScalarsResponse>, Status> {
+        let mut res: data::ListScalarsResponse = Default::default();
+        let mut run: data::list_scalars_response::RunEntry = Default::default();
+        run.run_name = "train".to_string();
+        run.tags.push(data::list_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            metadata: Some(data::ScalarMetadata {
+                max_step: 5,
+                max_wall_time: FAKE_START_TIME + 5.0,
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn read_scalars(
+        &self,
+        _request: Request<data::ReadScalarsRequest>,
+    ) -> Result<Response<data::ReadScalarsResponse>, Status> {
+        let mut res: data::ReadScalarsResponse = Default::default();
+        let mut run: data::read_scalars_response::RunEntry = Default::default();
+        run.run_name = "train".to_string();
+        run.tags.push(data::read_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            data: Some(data::ScalarData {
+                step: vec![0, 1, 2, 3, 4],
+                wall_time: (0..=4)
+                    .map(|i| FAKE_START_TIME + ((i + 1) as f64))
+                    .collect(),
+                value: vec![0.1, 0.5, 0.8, 0.9, 0.95],
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn list_tensors(
+        &self,
+        _request: Request<data::ListTensorsRequest>,
+    ) -> Result<Response<data::ListTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_tensors(
+        &self,
+        _request: Request<data::ReadTensorsRequest>,
+    ) -> Result<Response<data::ReadTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn list_blob_sequences(
+        &self,
+        _request: Request<data::ListBlobSequencesRequest>,
+    ) -> Result<Response<data::ListBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_blob_sequences(
+        &self,
+        _request: Request<data::ReadBlobSequencesRequest>,
+    ) -> Result<Response<data::ReadBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    type ReadBlobStream =
+        Pin<Box<dyn Stream<Item = Result<data::ReadBlobResponse, Status>> + Send + Sync + 'static>>;
```

Comment:
Thanks for the explanation around this type (the explanation is longer than the PR itself!).  I'm glad that creating a stream is simpler.

iiuc, the Send + Sync + 'static trait bounds all come directly from the type in the generated code from Prost in tensorboard.pb.rs?

Reply:
Yes. The trait definition specifies

```rust
type ReadBlobStream: Stream<Item = Result<super::ReadBlobResponse, tonic::Status>>
    + Send
    + Sync
    + 'static;
```

which means that our trait implementation must provide a type
`ReadBlobStream` that implements the [`Stream`] trait with the given
`Item` and also meets the `Send + Sync + 'static` bounds.

[`Stream`]: https://docs.rs/futures-core/0.3.8/futures_core/stream/trait.Stream.html